robot/client/client.go

```diff
@@ -288,6 +295,39 @@ func New(ctx context.Context, address string, clientLogger logging.ZapCompatible
 		return nil, err
 	}
 
+	// If running in a testing environment, wait for machine to report a state of
+	// running. We often establish connections in tests and expect resources to
+	// be immediately available once the web service has started; resources will
+	// not be available when the machine is still initializing.
+	//
+	// It is expected that golang SDK users will handle lack of resource
+	// availability due to the machine being in an initializing state themselves.
+	//
+	// Allow this behavior to be turned off in some tests that specifically want
+	// to examine the behavior of a machine in an initializing state through the
+	// use of an environment variable.
+	if testing.Testing() && os.Getenv(DoNotWaitForRunningEnvVar) == "" {
+		for {
+			if ctx.Err() != nil {
+				return nil, multierr.Combine(ctx.Err(), rc.conn.Close())
+			}
+
+			mStatus, err := rc.MachineStatus(ctx)
+			if err != nil {
+				// Allow for MachineStatus to not be injected/implemented in some tests.
+				if status.Code(err) == codes.Unimplemented {
+					break
+				}
+				return nil, multierr.Combine(err, rc.conn.Close())
```

Comment:
I'm going to recommend ignoring the error from `rc.conn.Close()` and just returning the `err` from `rc.MachineStatus` instead.

I've been undoing this pattern as I've come across it. Especially when correctness doesn't require `Close` to be called. See: the webrtc connection bugs where a connection succeeded, but we disconnected from the peer because `SendDone`[1] to the signaling server failed.

[1] I claim `SendDone` is analogous to `Close`.

Reply:
Gotcha that sounds good to me; thanks for pointing it out.